srtcore/api.cpp

```diff
@@ -2617,21 +2617,33 @@ void CUDTUnited::updateMux(
    }
    catch (CUDTException& e)
    {
-      m.m_pChannel->close();
-      delete m.m_pChannel;
-      throw;
+       updateMux_exceptionFallback(m);
+       throw;
    }
    catch (...)
    {
-      m.m_pChannel->close();
-      delete m.m_pChannel;
-      throw CUDTException(MJ_SYSTEMRES, MN_MEMORY, 0);
+       updateMux_exceptionFallback(m);
+       throw CUDTException(MJ_SYSTEMRES, MN_MEMORY, 0);
    }
 
    HLOGF(mglog.Debug, 
       "creating new multiplexer for port %i\n", m.m_iPort);
 }
 
+void CUDTUnited::updateMux_exceptionFallback(const CMultiplexer& m)
+{
+    // Reverse order of the assigned 
+    delete m.m_pRcvQueue;
+    delete m.m_pSndQueue;
+    delete m.m_pTimer;
+
+    if (m.m_pChannel)
+    {
+        m.m_pChannel->close();
+        delete m.m_pChannel;
+    }
+}
```

Comment:
Why not moving this function to the destructor of `CMultiplexer`?

Reply:
Because destructor will be called in every place where the object is deleted, including intermediate things happening during copying the object, at least formally. You'd need also to provide a copy constructor, with possibly some kind of poor C++03 move semantics (those from the `std::auto_ptr`).

Or, in short, not worth a shot.

Reply:
Still, this function looks more like a member of `CMultiplexer`, that a member of `CUDTUnited`.
Maybe move to `CMultiplexer::destroy()` or some similar name.

Another thought, `CMultiplexer::release()` to just set pointers to NULL without delete, and `~CMultiplexer()` destructor.

Reply:
I'd still leave the destructor behind, as it wouldn't be a good idea to let it run at the destruction time - but the idea to make this function a special method of `CMultiplexer` is very good.